### DISCUSSION TITLE
Dev Branch 2.2 Log Part 2

### DISCUSSION THREAD

**Discussion by MXKhronos** [HelixNebulaStudio/RiseOfTheDead#756] [General]
# Dev Branch 2.2 Log Part 2
Another dev branch dev log to talk about all the stuff in development. This one will be a lot more technical too because it relates to the overall progress for the Revive Engine itself. (fyi: Revive Engine is the engine that RotD is built on)

Initially RotD's content and the engine is completely tangled up, but over time, I've been separating the content from the engine itself. Content such as RotD missions, cutscenes, items, etc RotD related, separated from the tool mechanics, equipment systems. This is for Revive Engine to be reused in other projects such as Almonds. Systems like the tools system, storage system can be reused. To the players, this won't be any different. It also makes maintenance easier on my end.

## Completed

- **Engine Changes**
In live branch, all RotD mission content is now separated from the engine, along with the newly added lazy loading system that only loads assets e.g. cutscenes and sounds, when necessary, reducing memory usage on the client side.

    Last update on dev branch, all item modifiers have been migrated out of the engine too, e.g. damage mods, elemental mod mechanics, clothing mods, etc.. (Just means they are no longer loaded in Almonds, functionally it will be the same in RotD)

- **Npc Wield Customizations**
With the new customization system, the data that stores the customizations are [different](https://github.com/HelixNebulaStudio/RiseOfTheDead/blob/main/Source/Ref/Customizations.md). Npc's wield are not transitioned to use the new customizations system. The old system's removed hence their(Michael, Patrick) weapons customizations will be remade.
Bandits now has a stockless ak too as a neat touch.

- **New Item Modifier System**
Previously, adding a new mod mechanics e.g. dual shell, required modifying the engine's weapon handler script. This means everytime I add a new weapon mod with new mechanics, I have to extend the weapon handler script even more, currently at `2427` lines of code, it needed a new way of streamlining adding new mod mechanics.

    The new item modifier system is added to help with that, `skullburstmod` adds raw rpm to the weapon based on it's charge, it can easily hook into different parts of the weapon handler script to do so.
`WeaponHandler` checks what item modifiers are in the weapon when fired and checks if the item modifier has a OnBulletHit method, if so executes the function. The `skullburstmod`'s OnBulletHit function gets called and checks if it's a headshot -> adds charges -> calculate raw rpm to add based on mod upgraded's stats -> `WeaponHandler` adds the additional rawrpm to the next shot.

    The new item modifier system also hooks into the weapon's status hud system to display the mod statuses for `skullburstmod` and `frenzydamagemod`. It also supports displaying hotkeys to bring back new activatable mods.

    The new system makes it more convenient to implement new mod mechanics and continue to phase away old mod designs that passively gives plain stats for more unique designs that gives and takes.
  

- **Frenzy Damage Mod**
This is a new type of damage mod that increases your damage based on the damage you taken. Taking enemy damage charges up the mod. It used to be based on the % of health you're missing, but that concept isn't quite fun in practice, having to maintain a low health in order to maximize damage, even with max stacked warmonger's.
So now it's based on the initial damage you take before armor and other reductions (tire armor passive). More feedback needed.

 ## In Development

- **Ammo Pouch**
Proposed `ammopouch` functionality: _now act as a portable ammo refill station with higher ammo cost than shop_
Similar to GPS, after a single use, it'll be on cooldown, during the cooldown, using it again will cost significantly higher.

- **Hard Survivals**
Removal of 5 waves survival and endless will be default. 5 Waves mode was originally intended for some mission purposes but that has been scraped.
In Hard Survivals, enemy levels are doubled every 3 waves. So Wave 12 to 15 would be against enemies  Wave 12: lvl16s, Wave 13:~lvl21s, Wave14:~lvl26s, Wave15: lvl32s (Subject to Change)
Rewards items are dropped instead reward crates. These reward items are dropped every 3 waves in a defined sequence, e.g. Wave 3: `medkit`, Wave 6: `skullburstmod`, Wave 9: `boomerang`, etc..

- **Rechamber 1216**
New tier 5 shotgun, obtainable in hard survivals.

- **Uzi**
New tier 5 smg, obtainable in hard survivals.

- **~~Musket~~Grand Garand**
New tier 5 sniper, obtainable in hard survivals.

- **Nekron's Gambit**
A lot has been added to Nekron's Gambit since last introduction to the new game mode.

   **Pawns**
    Added pawns which pushes the lane. They will fight enemy pawns with melee and range. Each pawn wave consist of 1 melee pawn, 2 range pawn and a medic pawn. Medic pawn heals all allies within radius.

   **Sols**
    Currency and farming has been added. You earn **sols** by supporting your pawns kill enemy pawns. Enemy pawns dying in radius grants you sols, sol orbs that floats up can be shot for bonus sols. You can shoot both red and green sol orbs. Shooting allied sol orbs denies enemies from obtaining the bonus sol.

    **Scoreboard**
    Added scoreboard, currently shows player total sols, kills, death, and assist. Scoreboard can be opened by pressing `Tab`, rebindable in settings under Nekron's Gambit category.

   **Shop**
    Added WIP shop, open by pressing B. Shop contains all kinds of item modifiers. Weapon mods, Character mods, and Flex mods. 
    Weapon mods will add modifiers that increases stats, add passives or even active abilities.
    Character mods will add modifiers for character improvements in terms of health, armor, movement, utility.
    Flex mods are modifiers that gives you unique abilities such as throwing grenades etc..
    
    **Inventory**
    Inventory is also separated in 3 categories, Weapon(Red), Character(Green) and Flex(Blue).
    Weapon slots can only store weapon mods. Character slots for character mods.
    Flex slots can store any type of mods, so theoretically, you can have up to 8 weapon mods.

   **Pvp / Coop Vs AI**
   This game mode is intended to be pvp, but I would hope to add a Coop Vs AI mode in order to learn the gamemode / not participate in the intense pvp. This hope will be quite hard to work out to introduce NPC that can take a player's place.

    **King**
    Like any MOBA, there will be a throne to destroy in order to win. A King will take its place here as the final target.

![image](https://github.com/user-attachments/assets/c699b4a9-2f55-4192-aea2-b68cff16c24c)

- **Frostivus 2024**
There isn't anything planned for this december yet. Open to suggestions but due to time budgets. I'm mostly expecting a few new cosmetics and a new grindable badge level.

- **Cargo Ship Coop**
That will likely be resumed after the december update. 

## Conclusion
Quite a few things to work on. Still none of these are prioritized, they are just currently in development, and could be dropped or picked up at any moment.

Thanks for reading the dev log. If you have feedback or questions, feel free to ask, but since a lot is still in concept limbo, I might not be able to give a definite answer.
~Khronos

**Comment by JustDorien** [comment]
Peak as planned, I love this update so far, I want to see musket in the game tho, also make MASUS AND RIDE ON MOTHENA OBTAINABLEEEEEEEE, I also suggest making a whole battlepass so people won't be bored, but I would add exchanging coal for levels since people hate when it's only focused on doing missions, 5-6 new stuff would be cool and old stuff like 2020 presents, 2022 presents maybe as a reward, santa hat etc.

**Comment by MrPigZ** [comment]
Holly Jolly Time! 🎅

I would  like to suggest an event pass for christmas. Levels would be earned by trading 100 coal for 1 level. This would mean 1 post reward every 500 coal.
The main idea is based on time constraint which means that new never seen before way to grind 2024 presents won't be made. So, why not utilize all of the old stuff.

Treelum and Workshop
I would suggest replacing the present rewards with coal. So that players can choose to either buy the present from Klaws or spend it on event pass.
The pass levels can be grinded by killing zombies for coal drops, Treelum, Workshop, Railways Raid, mining coal, and the mission board. As an extra, I would like it if all game mode reward crates gave some coal so that players aren’t locked into doing something specific. (This sounds familiar 🤔)

Of course for christmas we need 2024 present. 🎁

What would the rewards for the pass be?
The free tiers could give:
The new skins so that they aren't behind rng wall 
2024 presents
Polish
Xp drinks
Tome of Tweaks
1 tile could give fireworks because new year celebration
Gold to buy the pass
Resources or christmas food as filler reward

Premium rewards could give:
Tome of Tweaks
2019 presents
2020 presents
2021 presents
Present 2022
Post rewards should give:
Custom colours 🤤
2019 presents
2020 presents
2021 presents
2022 presents
Ultra Omega Super Rare 2023 present
2024 presents
Coal in various amounts. For example unlucky roll where you get 30 and lucky roll where you get 750. If possible.
A suits skin (keeping the tradition)
Toxic waste containers

Why are there old gifts in post rewards? More gifts per coal is the name of the idea. You can buy a gift you want from Klaws, or you can gamble your coal and get more gifts.
I have no ideas for the Token Shop, but there will be lots of unlocks to flood the market so a token sink could be nice.


Also rising Klaws trade prices are somewhat uh. I dont think anyone want buy 2019 present for 1000 coal.

**Comment by MrGarrettFerret** [comment]
I think MrPigZ's idea for Frostivus 2024 is really good and I would like to see it implemented. I would suggest adding things like metal polish, xp drinks, and tomes of tweaks to either the post rewards or 2024 present droptable as they are in quite high demand atm.

I'd like to suggest a skin for Frostivus 2024 called "Wrapped". It'd available for all weapons. The skin itself would be similar to the 2022 xmas present: the weapon is wrapped in wrapping paper and has a bow tie on top of it. You could have the color values of the wrapping paper and bow be randomized (like Custom Colors) for an extra bit of collectability. Maybe there could be a few different variants of the wrapping paper to go along with it. Image below is a proof of concept of what it could look like
![wrappedskinconcept](https://github.com/user-attachments/assets/450bae96-145f-4baf-807c-047145053f64)

**Comment by Leroy8888** [comment]
Which wave is the player expected to reach with the new corrupted survival mode?

**Reply by MXKhronos** [reply]
Current the wave target for the rewards are arbitrary.
Once I finish the systems, I'll do some playtest to determine the numbers for the waves to level ratio, reward waves, and enemy tweaks. I've heard that enemies' movespeed is uncapped which is a problem and etc.
Also disable/adjust/new wave objectives.
